---
title: "NetBoxR Tutorial"
output:
  BiocStyle::html_document:
    toc: yes
    toc_float: false
  BiocStyle:pdf_document: default  
vignette: >
    %\VignetteIndexEntry{NetBoxR Tutorial}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
---

```{r knitrSetup, include=FALSE}
knitr::opts_chunk$set(out.extra='style="display:block; margin: auto"', fig.align="center", fig.width=12, fig.height=12, tidy=TRUE)
```



# Overview

The **netboxr** package composes a number of functions to retrive and process 
genetic data from large-scale genomics projects (e.g. TCGA projects) including 
from mutations, copy number alterations, gene expression and DNA methylation. 
The netboxr package implements NetBox algorithm in R package. NetBox algorithm 
integrates genetic alterations with literature-curated pathway knowledge to 
identify pathway modules in cancer. NetBox algorithm uses (1) global network 
null model and (2) local network null model to access the statistic significance 
of the discovered pathway modules.

# Basics
## Installation



# References

* Cerami E, Demir E, Schultz N, Taylor BS, Sander C (2010) Automated Network Analysis Identifies Core Pathways in Glioblastoma. PLoS ONE 5(2): e8918. doi:10.1371/journal.pone.0008918
* Cerami EG, Gross BE, Demir E, Rodchenkov I, Babur O, Anwar N, Schultz N, Bader GD, Sander C. Pathway Commons, a web resource for biological pathway data. Nucleic Acids Res. 2011 Jan;39(Database issue):D685-90. doi:10.1093/nar/gkq1039. Epub 2010 Nov 10.

# Session Information

```{r sessionInfo}
sessionInfo()
```
